src/coreclr/tools/Common/TypeSystem/IL/Stubs/AsyncThunks.cs

```diff
@@ -1,37 +1,257 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using Internal.TypeSystem;
 
 namespace Internal.IL.Stubs
 {
     public static class AsyncThunkILEmitter
     {
+        // The emitted code matches method EmitTaskReturningThunk in CoreCLR VM.
         public static MethodIL EmitTaskReturningThunk(MethodDesc taskReturningMethod, MethodDesc asyncMethod)
         {
             TypeSystemContext context = taskReturningMethod.Context;
 
             var emitter = new ILEmitter();
             var codestream = emitter.NewCodeStream();
 
-            // TODO: match EmitTaskReturningThunk in CoreCLR VM
-
             MethodSignature sig = asyncMethod.Signature;
-            int numParams = (sig.IsStatic || sig.IsExplicitThis) ? sig.Length : sig.Length + 1;
-            for (int i = 0; i < numParams; i++)
-                codestream.EmitLdArg(i);
+            TypeDesc returnType = sig.ReturnType;
+
+            MetadataType md = taskReturningMethod.Signature.ReturnType as MetadataType;
+            ReadOnlySpan<byte> name = md.Name;
+            bool isValueTask = name.SequenceEqual("ValueTask"u8) || name.SequenceEqual("ValueTask`1"u8);
```

Comment:
I'd use the same shortcut as the VM for consistency.

```suggestion
            bool isValueTask = returnType.IsValueType;
```